src/main/java/org/opensearch/security/user/User.java

```diff
@@ -272,15 +272,16 @@ public synchronized final Map<String, String> getCustomAttributesMap() {
         return attributes;
     }
 
-    public final void addSecurityRoles(final Collection<String> securityRoles) {
-        if (securityRoles != null && this.securityRoles != null) {
-            this.securityRoles.addAll(securityRoles);
+    public final void setSecurityRoles(final Collection<String> securityRoles) {
+        if(securityRoles != null) {
+            synchronized(this.securityRoles) {
+                this.securityRoles.clear();
+                this.securityRoles.addAll(securityRoles);
+            }
         }
     }
 
     public final Set<String> getSecurityRoles() {
-        return this.securityRoles == null
-            ? Collections.synchronizedSet(Collections.emptySet())
-            : Collections.unmodifiableSet(this.securityRoles);
+        return this.securityRoles == null ? Collections.synchronizedSet(Collections.emptySet()) : ImmutableSet.copyOf(this.securityRoles);
```

Comment:
I think it is okay we keep this change from the last commit.

Reply:
This should not be returning a new copy every time its called. If `securityRoles` is only set once, then there is no risk of concurrent modification. `Collections.unmodifiableSet(...)` would be better used here since it does not return a copy.